samza-sql/src/main/java/org/apache/samza/sql/avro/AvroRelConverter.java

```diff
@@ -164,25 +170,69 @@ public SamzaSqlRelMessage convertToRelMessage(KV<Object, Object> samzaMessage) {
     return new SamzaSqlRelMessage(samzaMessage.getKey(), fieldNames, values);
   }
 
+  private SamzaSqlRelRecord convertToRelRecord(GenericData.Record avroRecord) {
+    List<Object> values = new ArrayList<>();
+    List<String> fieldNames = new ArrayList<>();
+    if (avroRecord != null) {
+      fieldNames.addAll(avroRecord.getSchema().getFields()
+          .stream()
+          .map(Schema.Field::name)
+          .collect(Collectors.toList()));
+      values.addAll(avroRecord.getSchema().getFields()
+          .stream()
+          .map(x -> {
+            Object obj = avroRecord.get(x.name());
+            if (obj != null && obj instanceof GenericData.Record) {
+              return convertToRelRecord((GenericData.Record) obj);
+            }
+            return obj;
+          })
+          .collect(Collectors.toList()));
+    } else {
+      String msg = "Avro Record is null";
+      LOG.error(msg);
+      throw new SamzaException(msg);
+    }
+
+    return new SamzaSqlRelRecord(fieldNames, values);
+  }
+
+  /**
+   * Convert the nested relational message to the output message.
+   */
   @Override
   public KV<Object, Object> convertToSamzaMessage(SamzaSqlRelMessage relMessage) {
     return convertToSamzaMessage(relMessage, this.avroSchema);
   }
 
   protected KV<Object, Object> convertToSamzaMessage(SamzaSqlRelMessage relMessage, Schema avroSchema) {
+    return new KV<>(relMessage.getKey(), convertToGenericRecord(relMessage.getSamzaSqlRelRecord(), avroSchema));
+  }
+
+  private GenericRecord convertToGenericRecord(SamzaSqlRelRecord relRecord, Schema avroSchema) {
     GenericRecord record = new GenericData.Record(avroSchema);
-    List<String> fieldNames = relMessage.getFieldNames();
-    List<Object> values = relMessage.getFieldValues();
+    List<String> fieldNames = relRecord.getFieldNames();
+    List<Object> values = relRecord.getFieldValues();
     for (int index = 0; index < fieldNames.size(); index++) {
       if (!fieldNames.get(index).equalsIgnoreCase(SamzaSqlRelMessage.KEY_NAME)) {
-        record.put(fieldNames.get(index), values.get(index));
+        Object obj = values.get(index);
+        String fieldName = fieldNames.get(index);
+        if (obj instanceof SamzaSqlRelRecord) {
```

Comment:
nit: do we need "obj != null && ..." ?

Reply:
null check does not seem to be required before instanceof - https://docs.oracle.com/javase/specs/jls/se7/html/jls-15.html#jls-15.20.2